rfcs/ion_templates.md

```diff
@@ -0,0 +1,472 @@
+* Templates have a minimum length of 2 (1 byte type descriptor, 1 byte template ID)
+* A template blank (`L`=0, `@_`) found outside of a template definition or template invocation is illegal and should be treated as an error.
+* Template invocations with an ID that cannot be resolved are illegal and should be treated as an error. 
+    * (This differs from `symbol`s with an ID that cannot be resolved. `symbol`s are values while `template`s are an encoding detail. If a `Reader` encounters an unresolvable `symbol`, it can still tell the user that it is positioned over a symbol and even provide the unresolvable ID number.)
+
+
+
+## Text encoding within s-expressions
+
+Because `@` is a valid operator character within an s-expression, template blanks in the text format must be annotated with the special `$ion_template` annotation. Per the spec, symbols starting with `$ion` are reserved and may not be used in user data.
+
+### Blanks in a template definition s-expression
+
+```js
+$ion_template_table::{
+  templates : [{ // template 0
+    ($ion_template::@_)  
+  }]
+}  
+```
+
+### Template invocation inside an s-expression
+
+```js
+// Invoke template 0 from within an s-expression.
+($ion_template::@0(123))
+// ((123))
+```
+
+
+
+## Restrictions
+
+* Ion templates perform simple expansions. They involve no branching, looping, or recursion. They are guaranteed to terminate.
+* A template always expands to a single, complete Ion value.
+* Template invocations can only appear in positions where an Ion value could appear. (e.g. You cannot use a template invocation as an annotation or field name: `@1("dollars")::21.95` is illegal.)
+* Template definitions can refer to other templates, but only those templates that have already been defined (templates with a lower ID) in order to cheaply avoid cycles.
+
+
+
+## Support for suppressing values
+
+When `@_` appears as a parameter in a template invocation, the corresponding value in the template (and associated field name, if applicable) will be skipped over by the `Reader`.
+
+```js
+$ion_template_table::{
+  templates: [{ // template 0
+    name: @_,
+    employeeId: @_,
+    occupation: @_
+  }]
+}
+@0("Zack", @_, "SDE") // Explicitly leave 'employeeId' blank with '@_'
+```
+
+expands to
+
+```js
+{
+  name: "Zack",
+  occupation: SDE
+}
+```
+
+This feature, paired with [support for extending containers](#support-for-extending-containers) below, allows users to leverage existing template definitions to encode complex values rather than having to define a collection of very similar templates.
+
+
+## Support for extending containers
+
+If a template is a container type, invocations can provide enough values to fill in the template's blanks and then an extra parameter: a container of the same type as the template container. Values from this extra container will be appended to the end of the expanded container.
```

Comment:
Ah even more variation of the case of extra params provided. We should really be explicit about it if we are to rely on these variations and spell out all cases.